server/store/types/types.go

```diff
@@ -654,6 +664,75 @@ func (o AccessMode) Delta(n AccessMode) string {
 	return added + removed
 }
 
+// Applies the acs delta to AccessMode.
+// Delta is in the same format as generated by AccessMode.Delta.
+// E.g. JPRA.ApplyDelta(-PR+W) -> JWA.
+func (m *AccessMode) ApplyDelta(delta string) error {
+	if delta == "N" {
+		// No updates.
+		return nil
+	}
+	if delta == "" || (delta[0] != '+' && delta[0] != '-') {
+		return errors.New("AccessMode: invalid acs delta string '" + delta + "'")
+	}
+	var actions, acsDeltas []string
+	acceptAction := true
+	j := -1
+	for i := 0; i < len(delta); i++ {
+		switch delta[i] {
+		case '+':
+			fallthrough
+		case '-':
+			if acceptAction {
+				actions = append(actions, delta[i:i+1])
+				if j != -1 {
+					// [j..i) is the last acs delta
+					acsDeltas = append(acsDeltas, delta[j:i])
+				}
+				j = -1
+				acceptAction = false
+			} else {
+				return errors.New("Invalid acs delta string: repetitive or missing action '" + delta + "'")
+			}
+		default:
+			acceptAction = true
+			if j == -1 {
+				// Start of the new acs delta.
+				j = i
+			}
+		}
+	}
+	if j != -1 {
+		// Last piece.
+		acsDeltas = append(acsDeltas, delta[j:len(delta)])
+	}
+	if len(actions) != len(acsDeltas) {
+		return errors.New("Malformed acs delta string: '" + delta + "'")
+	}
+	m0 := *m
+	// Apply updates one by one.
+	for i := 0; i < len(actions); i++ {
+		if upd, err := DecodeAcs([]byte(acsDeltas[i])); err == nil {
+			switch actions[i] {
+			case "+":
+				if upd != ModeUnset {
+					m0 |= upd & ModeBitmask
+				}
+			case "-":
+				if upd != ModeUnset {
+					m0 &^= upd & ModeBitmask
+				}
+			default:
+				break
+			}
+		} else {
+			return err
+		}
+	}
+	*m = m0
+	return nil
+}
```

Comment:
This seems a bit too complex. I would rewrite it like this: https://play.golang.org/p/Y9ZFU-Dwn5G

Reply:
Cool. Changed.